Add support for incremental authorization

andyfmiller opened the pull request:
Fix for #1221 

There are two major parts to this fix:
1. Modify [ShouldRequestAuthorizationCode](https://github.com/google/google-api-dotnet-client/blob/master/Src/Support/Google.Apis.Auth/OAuth2/Web/AuthorizationCodeWebApp.cs#L128-L139) to detect a valid incremental authorization code.
2. Add a Flow.LoginHint property to simplify the user experience when requesting incremental authorization.

This PR also includes a Google.Apis.Auth.AspNetCore library since I want to use incremental authorization in an AspNetCore app.

There is a sample app here: https://github.com/andyfmiller/google-incremental-auth-sample

andyfmiller commented:
Added alternate email.

andyfmiller commented:
Comment for force googlebot check for CLA.

andyfmiller commented:
My email addresses in my CLA and here are in sync, but googlebot does not comment again. What else should I do?

chrisdunelm commented:
Thanks for this contribution :)
I'm not sure what's going on with the CLA. I'll have a proper look this week.

Please can you remove the ASP.NET Core MVC commit from this PR. I'd like to keep the two issues separate. If we decide to go down the route in PR #1109 then we'll work from that PR.
Once that first commit is removed (or create a new PR if that's simpler), then I'll have a proper look over the code.

andyfmiller commented:
Thank you for looking. I reverted the commit. Let me know if you'd prefer a new PR without any mention of the AspNetCore library.

chrisdunelm commented:
The problem with your CLA is that your CLA is signed as: `theafmiller@gmail.com`
But your commits are from: `andy@andyfmiller.com`

Please can you sign the CLA as `andy@andyfmiller.com`, then it should be OK.

chrisdunelm commented:
@andyfmiller Please can you take a look at the code I've added to this PR, especially the integration test, to confirm it covers the incremental auth case correctly. Thanks.

andyfmiller commented:
Comment to force CLA check.

andyfmiller commented:
@chrisdunelm The code you added definitely improves this PR, especially the integration test. Thank you!

chrisdunelm commented:
@andyfmiller I've come across a problem with this PR. You may have some suggestions on how best to resolve it.

`AuthorizationCodeFlow` now contains the following code:

```C#
// If the new token scope is not filled in, then the scope granted is identical to the scope
// requested. Scope is used to help determine if an authorization token is required.
// See AuthorizationCodeWebApp.ShouldRequestAuthorizationCode(TokenResponse token).
if (string.IsNullOrEmpty(token.Scope))
{
    token.Scope = authorizationCodeTokenReq.Scope;
}
```

But with incremental auth, `authorizationCodeTokenReq.Scope` will not contain all the scopes supported by the token; because the scopes may be asked for just one at a time, so `authorizationCodeTokenReq.Scope` only contains the most recently asked-for scope.
This then means that the code in `ShouldRequestAuthorizationCode` that examines the existing scopes and the requested scopes, to see if an auth request should be made, does not work.
I expect this is why currently `ShouldRequestAuthorizationCode` always thinks an authorization code is required if `IncludeGrantedScopes = true`.

I have a few thoughts:

* Are the changes in `ShouldRequestAuthorizationCode` really necessary? The user's code can track which scopes they've already asked for, so we don't need to.
* Or perhaps there is a way to force the token server to return all the scopes, so the `token.Scope = authorizationCodeTokenReq.Scope' code in `AuthorizationCodeFlow` is never required?
* (There is also the `tokenInfo` endpoint to retrieve the current scopes from the auth server, but this is specifically documented as being for dev/debug use only, and would involve another network round-trip which I'd rather avoid.)

So I'm now thinking that incremental auth is essentially already supported as best we can (i.e. without this PR). The only required addition is the `LoginHint` change in this PR.
Please let me know what you think.

andyfmiller commented:
@chrisdunelm I think the changes are really necessary. I can't use incremental authorization without them. Let me explain...

I am writing an AspNetCore app, following the advice for [OAuth 2.0 Authentication and Authorization](https://developers.google.com/api-client-library/dotnet/guide/aaa_oauth) page.  That page has 3 models: Installed Apps, Web Apps, and Service accounts. I'm following the Web Apps model. For example, to sign in I'm using:
```
            var result =
                await new AuthorizationCodeMvcApp(this, new SignInAppFlowMetadata(ClientId, ClientSecret))
                    .AuthorizeAsync(cancellationToken)
                    .ConfigureAwait(false);

            if (result.Credential == null)
            {
                return Redirect(result.RedirectUri);
            }

            return RedirectToAction("Index");
```
Where ```SignInAppFlowMetadata``` has "email" and "profile" scopes. And to get a list of Google Classroom classes, I use this:
```
            var flowData = new ClassListAppFlowMetadata(ClientId, ClientSecret);
            if (!string.IsNullOrEmpty(ViewData["PersonEmail"]?.ToString()))
            {
                flowData.LoginHint = ViewData["PersonEmail"].ToString();
            }

            // Incremental authorization
            var result = await new AuthorizationCodeMvcApp(this, flowData)
                .AuthorizeAsync(cancellationToken)
                .ConfigureAwait(false);

            if (result.Credential == null)
            {
                return Redirect(result.RedirectUri);
            }

            // Call ClassroomService to get the list of classes
```
Where ```ClassListAppFlowMetadata``` has "https://www.googleapis.com/auth/classroom.courses.readonly" scope.

If I don't specify ```IncludeGrantedScopes``` on the second authorization, ```AuthorizeAsync``` will just return the current token which is not authorized for "https://www.googleapis.com/auth/classroom.courses.readonly". ```AuthorizeAsync``` ignores the fact that ```ClassListAppFlowMetadata``` has the additional scope defined. I need to force ```AuthorizeAsync``` to request a new authorization code.

The docs for [Incremental Authorization](https://developers.google.com/identity/protocols/OAuth2WebServer#incrementalAuth) say I should specify ```IncludeGrantedScopes```.

That creates the next problem: ```result.Credential``` is always null if the ```GoogleAuthorizationCodeFlow.IncludeGrantedScopes``` is true. That means I never get past the authorization. When I put a breakpoint in ```ExchangeCodeForToken``` I could see that I got a new authorization code, but it was not converted into a token.

It turns out the reason ```result.Credential``` is always null is that ```AuthorizationCodeInstalledApp.ShouldRequestAuthorizationCode``` calls ```GoogleAuthorizationCodeFlow.ShouldForceTokenRetrieval``` which **always** returns true if ```IncludeGrantedScopes``` is true.

So I need to convince ```AuthorizationCodeInstalledApp.ShouldRequestAuthorizationCode``` or ```GoogleAuthorizationCodeFlow.ShouldForceTokenRetrieval``` to return true when I need a new token, then return false after the new token has been generated.

> Or perhaps there is a way to force the token server to return all the scopes, so the ```token.Scope = authorizationCodeTokenReq.Scope``` code in ```AuthorizationCodeFlow``` is never required?

This is my favorite option if I continue to use the token scopes to identify the new token, but I'm afraid I have no idea how to go about this.

Can you think of another way to detect that the authorization code or token has changed **and** it has the incremental scopes? Is it valid to simply look for change in the access_token (assuming the authorization code was returned without error)?

andyfmiller commented:
@chrisdunelm I added some changes to the PR to address your concern,

> But with incremental auth, authorizationCodeTokenReq.Scope will not contain all the scopes supported by the token; because the scopes may be asked for just one at a time, so authorizationCodeTokenReq.Scope only contains the most recently asked-for scope.

Now the previously granted scopes are added to the token.Scope (but only when ```IncludeGrantedScopes``` is true).

I'm sorry for the messy set of commits. I missed changing some changed files, and added a file I did not want to push. Eventually I got it straight, and all the tests pass.

chrisdunelm commented:
@andyfmiller I understand your use-case; thanks for the explanation.

It looks to me like it should be possibly for you to get your desired behaviour without the changes in this PR. By keeping track of which scopes are already requested/authorized, you know whether you need to set `IncludeGrantedScopes` to true or not. If not setting it, then the auth call should include all scopes, not just the incremental scope(s).
I think this shouldn't be too much problem to implement.

However, I agree it would be better if this library just did all this for you, so you didn't have to worry about it in your code.
But this is tricky. Unfortunately we cannot use the implementation you've provided because it breaks backward compatibility on various classes/interfaces. The interface breakage is the most tiresome to deal with, as changing an interface in *any way* breaks backward compatibility, which we cannot do.
I will see if I can come up with alternative way to implement this without requiring the changes to `IAuthorizationCodeFlow`.
However, this isn't a very high priority work item, so I suggest you use the workaround for now. If I've made a mistake, and the workaround isn't possible, then please let me know.

(We use [semantic versioning](https://semver.org/), so backward incompatible changes are only acceptable in a major version upgrade; but we are not planning to do any major version updates of this library)

andyfmiller commented:
@chrisdunelm I understand. I'll close this PR and use the workaround.
